### PR TITLE
Show samples count in Case Set Selector for multi-study selection

### PR DESCRIPTION
**Pull request by kalletlak**
# What? Why?
Fixes 
1. https://github.com/cBioPortal/cbioportal/issues/4308
2. https://github.com/cBioPortal/cbioportal/issues/4337

### CODE REVIEW

File: src/shared/components/query/CaseSetSelector.tsx [hunk 1]
```diff
@@ -70,24 +70,37 @@ export default class CaseSetSelector extends QueryStoreComponent<{}, {}>
...
+						count = mutation_cna;
+					}
+				} else {
+					//for case set All
+					count = this.store.selectableSelectedStudies_totalSampleCount
+				}
+
+				//add profile only if it has samples
+				if (count > 0) {
+					acc.push($.extend({}, next, { name: `${next.name} (${count})` }))
```
Comment (alisman) [inline src/shared/components/query/CaseSetSelector.tsx]: we shouldn't use jQuery here. lets use Object.assign or lodash methdo

File: src/shared/cache/SampleListsInStudyCache.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    let promises = studyIds.map(studyId => client.getAllSampleListsInStudyUsingGET({
+        studyId: studyId,
+        projection: 'DETAILED'
+    }))
+
+    const sampleLists:SampleList[] = await Promise.all(promises).then((allData) => {
+        return _.flatMap(allData)
+    })
+    const sampleListsByStudy = _.groupBy(sampleLists, profile=>profile.studyId);
```
Comment (alisman) [inline src/shared/cache/SampleListsInStudyCache.ts]: sampleListsByStudyId

File: src/shared/cache/SampleListsInStudyCache.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    return studyIds.map(studyId=>{
+        const data = [(sampleListsByStudy[studyId] || [])];
```
Comment (alisman) [inline src/shared/cache/SampleListsInStudyCache.ts]: seems odd.  do you need the enclosing array?  why would there be an array with a single empty array as only item?

File: src/shared/cache/SampleListsInStudyCache.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        return {
+            data,
+            meta: studyId
```
Comment (alisman) [inline src/shared/cache/SampleListsInStudyCache.ts]: I would do { meta:  { studyId } } as  meta gives no info about what that property is

File: src/shared/cache/SampleListsInStudyCache.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import LazyMobXCache, {AugmentedData} from "../lib/LazyMobXCache";
```
Comment (alisman) [inline src/shared/cache/SampleListsInStudyCache.ts]: @adamabeshouse we should be using new cache, yes?

Comment (adamabeshouse) [reply src/shared/cache/SampleListsInStudyCache.ts]: yeah, definitely

File: src/shared/cache/SampleListsInStudyCache.ts (hunk 3 above)
Comment (kalletlak) [reply src/shared/cache/SampleListsInStudyCache.ts]: Its what `LazyMobXCache` require. we have implementation even in other cache implementations

File: src/shared/cache/SampleListsInStudyCache.ts (hunk 5 above)
Comment (kalletlak) [reply src/shared/cache/SampleListsInStudyCache.ts]: @alisman @adamabeshouse new cache? example?

Comment (kalletlak) [reply src/shared/cache/SampleListsInStudyCache.ts]: you mean `MobxPromiseCache`

Comment (adamabeshouse) [reply src/shared/cache/SampleListsInStudyCache.ts]: @kalletlak yes, let me know if you have questions about how to use it. I think you'll find its cleaner/easier

File: src/pages/resultsView/ResultsViewPage.tsx [hunk 6]
```diff
@@ -64,7 +64,7 @@ function initStore(queryStore: QueryStore) {
     // ultimate we will phase this out and this information will be stored in router etc.
     //const qSession:any = (window as any).QuerySession;
     var samplesSpecification:any = [];
-    if(_.includes(['all', '0', '1', '2'],serverVars.caseSetProperties.case_set_id)){
+    if(_.includes(['all', 'w_mut_cna', 'w_mut', 'w_cna'],serverVars.caseSetProperties.case_set_id)){
```
Comment (alisman) [inline src/pages/resultsView/ResultsViewPage.tsx]: a corresponding change was made in backend?

Comment (kalletlak) [reply src/pages/resultsView/ResultsViewPage.tsx]: Yes! there's a pr to it https://github.com/cBioPortal/cbioportal/pull/4341

File: src/shared/components/query/QueryStore.ts [hunk 7]
```diff
@@ -592,15 +607,71 @@ export class QueryStore
 		}
 	});
 
+	readonly profiledSamplesCount = remoteData<{ mutation: number, cna: number, mutation_cna: number }>({
+		await: () => [this.sampleListInSelectedStudies],
+		invoke: async () => {
+			const profiledSamples = await Promise.resolve(categorizedSamples(this.sampleListInSelectedStudies.result!));
```
Comment (alisman) [inline src/shared/components/query/QueryStore.ts]: it seems categorizedSamples is a synchronous function, so why do we need to wrap in a promise and then await it?

File: src/shared/components/query/CaseSetSelector.tsx [hunk 8]
```diff
@@ -70,24 +70,37 @@ export default class CaseSetSelector extends QueryStoreComponent<{}, {}>
...
+		let {mutation,cna,mutation_cna} = this.store.profiledSamplesCount.result;
 
 		//filter case lists based on molecular profiles
-		let filteredcustomCaseSets = CustomCaseSets.filter(s=>{
-			if(this.store.isVirtualStudyQuery){
-				if( s.dataType && 
-					(!mutation || !cna) && 
-					(s.dataType.mutation !== mutation || s.dataType.cna !== cna)){
-					return false
+		let filteredcustomCaseSets = _.reduce(CustomCaseSets, (acc: CustomCaseSet[], next) => {
```
Comment (alisman) [inline src/shared/components/query/CaseSetSelector.tsx]: this function should be in a util file and have a test.

File: src/shared/components/query/QueryStore.ts (hunk 7 above)
Comment (kalletlak) [reply src/shared/components/query/QueryStore.ts]: true, its not rquired. Forgot to refactor when changing from LazyMobXCache to MobxPromiseCache

File: src/shared/components/query/QueryStoreUtils.spec.ts [hunk 9]
```diff
@@ -128,4 +128,67 @@ describe('QueryStoreUtils', ()=>{
...
+            assert.deepEqual(categorizedSamples(mutationSampleLists), {mutation:{study1:{sample1:true},study2:{sample1:true}}, cna:{}, mutation_cna:{}});
+        });
+        it("returns correct categoried samples when only `all_cases_with_cna_data` sets are present", ()=>{
+            assert.deepEqual(categorizedSamples(cnaSampleLists), {mutation:{}, cna:{study1:{sample1:true}}, mutation_cna:{}});
+        });
+        it("returns correct categoried samples when only `all_cases_with_mutation_and_cna_data` sets are present", ()=>{
+            assert.deepEqual(categorizedSamples(mutationCnaSampleLists), {mutation:{}, cna:{}, mutation_cna:{study1:{sample1:true}}});
+        });
+        it("returns correct categoried samples when only sets are present", ()=>{
+            console.log(JSON.stringify(categorizedSamples([...mutationSampleLists,...cnaSampleLists,...mutationCnaSampleLists])))
```
Comment (alisman) [inline src/shared/components/query/QueryStoreUtils.spec.ts]: kill console

File: src/shared/components/query/CaseSetSelectorUtils.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+	{name: 'User-defined Case List', description: 'Specify your own case list', value: CaseSetId.custom, isdefault : true}
+]
+
+export function getFilteredCustomCaseSets(isVirtualStudy:boolean, profiledSamplesCount: { mutation: number; cna: number; mutation_cna: number; all: number}) {
+    return _.reduce(CustomCaseSets, (acc: CustomCaseSet[], next) => {
+        if (next.isdefault) {
+            acc.push(next)
+        } else if (isVirtualStudy) {
+            let count = 0;
+            switch(next.value) {
```
Comment (alisman) [inline src/shared/components/query/CaseSetSelectorUtils.ts]: you don't have to change this but wouldn't this be more simply written:
profiledSamplesCount.mutation.length || profiledSamplesCount.cna.length || ... 
No switch statement, no count variable.

File: src/shared/components/query/QueryStoreUtils.ts [hunk 11]
```diff
@@ -119,34 +120,88 @@ export function profileAvailability(molecularProfiles:MolecularProfile[]) {
...
+            
+            // check if the study in this virtual study is already in the selected studies list
+            // and only add the samples if its not already present
+            if(!_.includes(selectedPhysicalStudyIds,study.id)){
+                filteredMutationSamples[study.id] = filteredMutationSamples[study.id] || {};
+                filteredCnaSamples[study.id] = filteredCnaSamples[study.id] || {};
+                filteredMutationCnaSamples[study.id] = filteredMutationCnaSamples[study.id] || {};
+                filteredallSamples[study.id] = filteredallSamples[study.id] || {};
+
+                _.forEach(study.samples,sample=>{
```
Comment (alisman) [inline src/shared/components/query/QueryStoreUtils.ts]: this is kind of confusing, but it's not sample, it's sampleId.  We need to be really careful about what we name things.  I know it comes from VirtualStudy model, but that should be changed.

File: src/shared/components/query/CaseSetSelectorUtils.ts (hunk 10 above)
Comment (kalletlak) [reply src/shared/components/query/CaseSetSelectorUtils.ts]: Actually I need the `count` here. It is append to the sample list name at line 52.

File: src/shared/components/query/QueryStoreUtils.ts (hunk 11 above)
Comment (kalletlak) [reply src/shared/components/query/QueryStoreUtils.ts]: I'll change it

File: src/shared/components/query/CaseSetSelectorUtils.ts (hunk 10 above)
Comment (kalletlak) [reply src/shared/components/query/CaseSetSelectorUtils.ts]: `value` is of type enum
```
enum CaseSetId {
	"custom"       = '-1',
	"mutation_cna" = 'w_mut_cna',
	"mutation"     = 'w_mut',
	'cna'          = 'w_cna',
	'all'          = 'all'
}
```
Then I need to change `profiledSamplesCount` to `{w_mutation:number, w_cna:number, w_mutation_cna:number, all:number}`

**Pull request merged**